controller/object_store_controller.go

```
@@ -0,0 +1,1277 @@
+	}
+
+	dpl, store, err := osc.getOrCreateDeployment(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating deployment")
+	}
+
+	_, store, err = osc.getOrCreateService(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating service")
+	}
+
+	endpoints, store, err := osc.getOrCreateS3Endpoints(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating S3 ingresses")
+	}
+	osc.logger.Infof("object store %v has  %v S3 endpoint(s)", store.Name, len(endpoints))
+	// if there are no public endpoints, add the implicit cluster-internal one
+	if len(store.Status.Endpoints) == 0 {
+		store.Status.Endpoints = append(store.Status.Endpoints, fmt.Sprintf("%v.%v.svc", store.Name, osc.namespace))
+	}
+
+	if err := osc.checkVolume(vol); err != nil {
+		return nil
+	}
+
+	if err := osc.checkPV(pv); err != nil {
+		return nil
+	}
+
+	if err := osc.checkPVC(pvc); err != nil {
+		return nil
+	}
+
+	if err := osc.checkDeployment(dpl, store); err != nil {
+		return nil
+	}
+
+	logrus.Infof("object store %v is now running", store.Name)
+	store.Status.State = longhorn.ObjectStoreStateRunning
+	return nil
+}
+
+// This function does a short sanity check on the various resources that are
+// needed to operate the object stores. If any of them is found to be
+// unhealthy, the controller will transition the object store to "Error"
+// state, otherwise do nothing.
+func (osc *ObjectStoreController) handleRunning(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
+		logrus.Infof("object store %v is now stopping", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStopping
+		return nil
+	}
+
+	// TODO: return errors if we encounter them
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return errors.Wrapf(err, "failed to find deployment %v", store.Name)
+		}
```

Should this function continue execution if `err != nil && !datastore.ErrorIsNotFound(err)`

---

No, that's right. As described [here](https://github.com/longhorn/longhorn-manager/pull/2136#discussion_r1406671471), this should just return an error in any case. We'll later decide if it was an server-side API error or something else and depending on that update the object store status.